### ISSUE TITLE
Fix/drop AutocompleteInput

### ISSUE THREAD

**Issue by azjps** [bokeh/bokeh#5596]
[On linux chrome v55.0.2883.87, bokeh master] The menu revealed by AutocompleteInput does not have any css styling, it is just floating text which overlaps with other elements on the page. I assume at some point some jquery css had been excised causing this widget to break. With the current trend I assume that this widget should just get removed. (I'd find it useful though to have an example of a custom AutocompleteInput widget.)

**Comment by bryevdv** [comment]
I would support removing `Autocomplete` and adding a custom extension wrapping/showing some autocmoplete at the same time.

**Comment by mattpap** [comment]
This is the same story as with `DatePicker` widget (#5506).

**Comment by bryevdv** [comment]
True, tho before going do this route we need to find the robust solution to extensions.

**Comment by azjps** [comment]
> robust solution to extensions

So this is this something I should wait until someone else handles? At any rate as a temporary workaround for myself, I can get this to work as a custom extension by pointing `__css__` to jquery-ui from CDN.

**Comment by mattpap** [comment]
> True, tho before going down this route we need to find the robust solution to extensions.

Without actually creating extensions we won't figure out what's the robust approach.

**Comment by azjps** [comment]
I wrote a custom extension using [awesomplete (lightweight, no dependencies)](https://leaverou.github.io/awesomplete/), which I'd hope would be a good candidate for a replacement example, but weirdly the awesomplete bindings get deregistered when the completion menu is present and then focus is removed (say by clicking elsewhere). (I've been watching the binding's via chrome's `Element > Event Listeners` tab; in some bokeh models the bindings aren't even registered when the model is first rendered, even though the awesomplete bindings have been invoked.) I have to admit its driving me a bit crazy, especially with my lack of JS knowledge. This doesn't happen with any of the examples on the awesomplete website. Would one of you mind confirming this behavior for me? Here's the code for the plugin:

```python
# awesomplete_input.py
from bokeh.core.properties import List, String, Int, Bool
from bokeh.models.widgets import TextInput

class AwesompleteInput(TextInput):
    """ Single-line input widget with auto-completion. """

    __implementation__ = "awesomplete_input.coffee"
    __css__ = ["https://cdnjs.cloudflare.com/ajax/libs/awesomplete/1.1.1/awesomplete.css"]
    __javascript__ = ["https://cdnjs.cloudflare.com/ajax/libs/awesomplete/1.1.1/awesomplete.js"]

    completions = List(String, help="""
    A list of completion strings. This will be used to guide the
    user upon typing a substring of a desired value.
    """)

    min_chars = Int(default=2, help="""
    Minimum characteres the user has to type before the autocomplete
    popup shows up.
    """)

    max_items = Int(default=10, help="""
    Maximum number of suggestions to display.
    """)

    auto_first = Bool(default=False, help="""
    Whether the first element should be automatically selected
    """)
```

```coffeescript
# awesomplete_input.coffee
import {TextInput, TextInputView} from "models/widgets/text_input"
import * as p from "core/properties"


export class AwesompleteInputView extends TextInputView

  initialize: (options) ->
    @awesomplete = null
    super(options)
    # already called in super
    # @listenTo(@model, 'change', @render)
    # @render()

  render: () ->
    super()
    if @awesomplete == null
      # fixme, init-ing over here because render() is called in super() ..
      $input = @$el.find("input")
      $input.addClass("dropdown-input")
      @awesomplete = new Awesomplete($input[0])
    @awesomplete.list = @model.completions
    @awesomplete.minChars = @model.min_chars
    @awesomplete.maxItems = @model.max_items
    @awesomplete.autoFirst = @model.auto_first
    return @

export class AwesompleteInput extends TextInput
  type: "AwesompleteInput"
  default_view: AwesompleteInputView

  # defaults are from awesomplete
  @define {
      completions: [ p.Array, [] ]
      min_chars: [ p.Number, 2 ] 
      max_items: [ p.Number, 10 ]
      auto_first: [ p.Bool, false ]
    } 
```

```python
# main.py
from bokeh.layouts import layout
from bokeh.models.widgets import Div, Slider
from bokeh.plotting import curdoc

from awesomplete_input import AwesompleteInput

auto_input = AwesompleteInput(completions=["/Option/%d/" % i for i in range(20)])
div = Div(text="Random widget below dropdown menu")
slider = Slider(start=0, end=10, value=0, step=1, title="Slider")

curdoc().add_root(layout([[auto_input], [div], [slider]]))
```

**Comment by azjps** [comment]
I should have realized this, but the above question was just the same issue as #5597 where the DOM element gets destructed and re-rendered, in conjunction with the fact that `TextInput.on_change` is triggered by losing focus from the input box. I haven't dug to see why this decision was made (I assume there's a good reason to trade off the cost of re-rendering and having to avoid memory leaks, etc; maybe so that it is easier to listen to all model changes?).

I can submit a PR then removing autocomplete and replacing with either jquery or awesomplete.

**Comment by ttal** [comment]
@mattpap this may not be the best way to do this, but you can refresh awesomplete by initializing it on every render rather than only on null:

  render: () ->
    super()
    #if @awesomplete == null
      # fixme, init-ing over here because render() is called in super() ..
    $input = @$el.find("input")
    $input.addClass("dropdown-input")
    @awesomplete = new Awesomplete($input[0])
    @awesomplete.list = @model.completions
    @awesomplete.minChars = @model.min_chars
    @awesomplete.maxItems = @model.max_items
    @awesomplete.autoFirst = @model.auto_first
    return @

**Comment by lukauskas** [comment]
If anyone is still in favour of bringing this widget back, I've done some work on integration of bokeh with [`selectize.js`](https://selectize.github.io/selectize.js/) for my project, see:

https://github.com/lukauskas/bokeh-selectize

It currently is designed as an extension to bokeh, and sort of mimics `MultiSelect` API, but in theory could replace autocompleteinput, select and multiselect widgets all at once.
I am currently using it with `bokeh==0.12.6.dev4` with no problems.

Not sure how one would go integrating it with core `bokeh` though.

**Comment by mattpap** [comment]
This was fixed in PR #6335.
